### DISCUSSION TITLE
fetch post not working

### DISCUSSION THREAD

**Discussion by ashok777** [jupyterlite/jupyterlite#412] [Q&A] [answered]
I am trying to execute a POST request using jupyterlite fetch. However, the request that goes out is a Http Get (as seen in the browser's developer tools network panel). Wondering how I should provide the payload for the POST to work. Here's the code

`url ="https://httpbin.org/post"
hdrs = {'accept':'application/json'}

response = await fetch("https://httpbin.org/post", {'headers':hdrs,'body':'{}','method':'POST'});

res = await response.text();  
print(res)`

**Comment by jtpio** [comment, marked as answer]
@ashok777 normally this code snippet should work:

```python
import json
from pyodide import to_js
from IPython.display import JSON
from js import Object, fetch

resp = await fetch('https://httpbin.org/post',
  method="POST",
  body=json.dumps({ "some" : "json" }),
  credentials="same-origin",
  headers=Object.fromEntries(to_js({ "Content-Type": "application/json" })),
)
res = await resp.text()
JSON(json.loads(res))
```

![image](https://user-images.githubusercontent.com/591645/140499753-f8d0874d-45e7-4b98-9f3b-578894ec0101.png)


The difference is the extra parameters being passed as keyword arguments.

**Reply by ashok777** [reply]
Thanks @jtpio. The code snippet you provided works. However, I am finding that fetch operations (both GET and POST) work only when the protocol is https, and fails for http.

**Comment by jtpio** [comment]
The Pyodide documentation has some information about this: https://pyodide.org/en/stable/usage/faq.html#how-can-i-use-fetch-with-optional-arguments-from-python

**Comment by bollwyvl** [comment]
If serving jupyterlite from https, it's possible the browser might not be
able to fetch http resources... Or may require some additional settings.
What exact error are you seeing, and where did it appear? Do you control
the other end of the pipe? Perhaps there is a proxy you could use, e.g.
archive.org?

**Reply by ashok777** [reply]
It appears that the problem is within fetch. I tried multiple http sites and in each case the exception was the same : 

JsException: TypeError: Failed to fetch

One of the sites I tried : http://eu.httpbin.org/

I am able to successfully access the site from the browser, or even using Requests in a jupyterlab notebook. The problem is seen only on fetch in jupyterlite. The failing call: await fetch('http://eu.httpbin.org/'). As I indicated before https calls are working reliably.

**Reply by rth** [reply]
If you try,
```js
await fetch('http://eu.httpbin.org')
```
in the browser JS console, you would get
```
Referrer Policy: Less restricted policies, including ‘no-referrer-when-downgrade’, ‘origin-when-cross-origin’ and ‘unsafe-url’, will be ignored soon for the cross-site request: http://eu.httpbin.org/ debugger eval code:2:21
Blocked loading mixed active content “http://eu.httpbin.org/”
```
It's a browser security feature https://support.mozilla.org/en-US/kb/mixed-content-blocking-firefox and the behavior is expected.

You should see the same error in the JS console if you run fetch from Python.

**Reply by ashok777** [reply]
@bollwyvl @rth : Thank you both for chiming in. I've got fetch/http working now. It turned out that CORS was what was causing the issue in my case. By having both the jupyterlite notebook service request and the fetch http domain point to the same origin the problem did not occur. I realize that mixed content from http/https can also cause the type of problem seen as pointed out by both of you. I also realize everything has got to be https and that is where I'll be headed soon.
